dora/core/common/src/main/java/alluxio/collections/IndexedSetTrie.java

```diff
@@ -0,0 +1,374 @@
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.collections;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+
+import java.util.AbstractSet;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A set of objects that are indexed and thus can be queried by specific fields of the object.
+ * Different {@link IndexedSetTrie} instances may specify different fields to index. The field type
+ * must be comparable. The field value must not be changed after an object is added to the set,
+ * otherwise, behavior for all operations is not specified.
+ *
+ * If concurrent adds or removes for objects which are equivalent, but not the same exact object,
+ * the behavior is undefined. Therefore, do not add or remove "clones" objects in the
+ * {@link IndexedSetTrie}.
+ *
+ * <p>
+ * Example usage:
+ *
+ * We have a set of puppies:
+ *
+ * <pre>
+ *   class Puppy {
+ *     private final String mName;
+ *     private final long mId;
+ *
+ *     public Puppy(String name, long id) {
+ *       mName = name;
+ *       mId = id;
+ *     }
+ *
+ *     public String name() {
+ *       return mName;
+ *     }
+ *
+ *     public long id() {
+ *       return mId;
+ *     }
+ *   }
+ * </pre>
+ *
+ * We want to be able to retrieve the set of puppies via a puppy's id or name, one way is to have
+ * two maps like {@code Map<String, Puppy> nameToPuppy} and {@code Map<Long, Puppy> idToPuppy},
+ * another way is to use a single instance of {@link IndexedSetTrie}!
+ *
+ * First, define the fields to be indexed:
+ * <pre>
+ *  IndexDefinition&lt;Puppy&gt; idIndex = new IndexDefinition&lt;Puppy&gt;(true) {
+ *    {@literal @Override}
+ *    Object getFieldValue(Puppy o) {
+ *      return o.id();
+ *    }
+ *  }
+ *
+ *  IndexDefinition&lt;Puppy&gt; nameIndex = new IndexDefinition&lt;Puppy&gt;(true) {
+ *    {@literal @Override}
+ *    Object getFieldValue(Puppy o) {
+ *      return o.name();
+ *    }
+ *  }
+ * </pre>
+ *
+ * Then create an {@link IndexedSetTrie} and add puppies:
+ * <pre>
+ *  IndexedSet&lt;Puppy&gt; puppies = new IndexedSet&lt;Puppy&gt;(idIndex, nameIndex);
+ *  puppies.add(new Puppy("sweet", 0));
+ *  puppies.add(new Puppy("heart", 1));
+ * </pre>
+ *
+ * Then retrieve the puppy named sweet:
+ * <pre>
+ *   Puppy sweet = puppies.getFirstByField(nameIndex, "sweet");
+ * </pre>
+ * and retrieve the puppy with id 1:
+ * <pre>
+ *   Puppy heart = puppies.getFirstByField(idIndex, 1L);
+ * </pre>
+ *
+ * @param <T> the type of object
+ */
+@ThreadSafe
+public class IndexedSetTrie<T> extends AbstractSet<T> {
```

Comment:
If the only methods you modified is the ctor and the `getByPrefix`, why not regarding `IndexedSetTrie` as a descedant of `IndexedSet`? Then you just need to change ctor and `gerByPrefix`, instead of copying all the methods.